compiler/sem/semstmts.nim

```diff
@@ -1659,76 +1659,90 @@ proc semCase(c: PContext, n: PNode; flags: TExprFlags): PNode =
     if skipTypes(caseTyp[0], abstractInst).kind in shouldChckCovered:
       chckCovered = true
   of tyFloat..tyFloat64, tyString:
-    # xxx: possible case statement macro bug, as it'll be skipped here
-    discard
+    discard "not all possible values have to be covered"
   else:
-    popCaseContext(c)
     closeScope(c)
-    result[0] = c.config.newError(n[0],
+    result[0] = c.config.newError(result[0],
                           PAstDiag(kind: adSemSelectorMustBeOfCertainTypes))
+    result = c.config.wrapError(result)
     return
+
+  pushCaseContext(c, result) # push the in-progress case context
   for i in 1..<n.len:
     setCaseContextIdx(c, i)
-    var x = n[i]
+    let x = n[i]
     when defined(nimsuggest):
       if c.config.ideCmd == ideSug and c.config.m.trackPos == x.info and caseTyp.kind == tyEnum:
         suggestEnum(c, x, caseTyp)
     case x.kind
     of nkOfBranch:
-      checkMinSonsLen(x, 2, c.config)
-      semCaseBranch(c, n, x, i, covered)
-      var last = x.len-1
-      x[last] = semExprBranchScope(c, x[last])
-      typ = commonType(c, typ, x[last])
+      let branch = semCaseBranch(c, selector, x, covered)
+      result[i] = branch
+      checkBranchForOverlap(c, result, i, result[i].len - 1)
+      # the branch node might be inspected from within the body, so make sure
+      # it is syntactically valid prior to the analysis
+      branch.add c.graph.emptyNode
+      branch[^1] = semExprBranchScope(c, x[^1])
     of nkElifBranch:
       chckCovered = false
       checkSonsLen(x, 2, c.config)
       openScope(c)
-      x[0] = forceBool(c, semExprWithType(c, x[0]))
-      x[1] = semExprBranch(c, x[1])
-      typ = commonType(c, typ, x[1])
+      let branch = shallowCopy(x)
+      branch[0] = forceBool(c, semExprWithType(c, x[0]))
+      branch[1] = semExprBranch(c, x[1])
       closeScope(c)
+      result[i] = branch
     of nkElse:
       checkSonsLen(x, 1, c.config)
-      x[0] = semExprBranchScope(c, x[0])
-      typ = commonType(c, typ, x[0])
-      if (chckCovered and covered == toCover(c, n[0].typ)) or hasElse:
+      let branch = shallowCopy(x)
+      branch[0] = semExprBranchScope(c, x[0])
+      result[i] = branch
+      if (chckCovered and covered == toCover(c, selector)) or hasElse:
         localReport(c.config, x.info, SemReport(kind: rsemUnreachableElse))
       hasElse = true
       chckCovered = false
     else:
```

Comment:
```suggestion
    of nkError:
      result[i] = x
    else:
```

Do we want to do this instead so `nkError` is nested and then wrapped and reported correctly later on?

Reply:
Thought about this again, and I think if we have an error then it should be illegal syntax unless it's already an illegal syntax error (avoid repeated wrapping). As for better reporting that should be handled in the illegal syntax error message generation code.

Ultimately, I think this means my suggestion is also not correct as it needs to check for the error kind in the node and then let it through if illegal syntax otherwise wrap it.

Thoughts?

Reply:
The only two reasons an `nkError` node can appear here are:
1. the error wrapping in `semCase` is faulty
2. a macro transplanted an error it somehow got hold of into the AST

Number 1 would be a compiler bug, and number 2 could also be considered one (macros shouldn't be able to get hold of `nkError` nodes, at least at the moment), so I think reporting an illegal syntax error (which raises an exception) is a reasonable thing to happen.

Alternatively, `internalError` could be used, though it'd be inconsistent with all other analysis (e.g., `semIf`, `semTry`, etc.), where `nkError`s in the input are treated as illegal syntax.

Reply:
Makes sense, I suspect the second option is quite possible, but the current approach is reasonable.